### ISSUE TITLE
buf should check if the file is actually protobuf file

### ISSUE THREAD

**Issue by aw-was-here** [bufbuild/buf#107]
During testing of a new (private) repository, the directory structure is:

```
/dir
- Makefile.proto
- a/
  - a.proto
- b/
  - b.proto
- c/
  - c.proto
```

The Makefile.proto file is not a protobuf file but actually a Makefile.  While that's clearly a problem,  buf throws a SEGV and this stack trace:

```
$ buf check lint
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x99371b]

goroutine 18 [running]:
github.com/jhump/protoreflect/desc/protoparse.(*basicCompositeNode).start(...)
github.com/jhump/protoreflect@v1.7.1-0.20200706205124-8d2d934ce565/desc/protoparse/ast.go:162
github.com/jhump/protoreflect/desc/protoparse.(*parseResult).createFileDescriptor(0xc00039c0f0, 0xc0002b6c30, 0xe, 0xc00038c180)
github.com/jhump/protoreflect@v1.7.1-0.20200706205124-8d2d934ce565/desc/protoparse/descriptor_protos.go:35 +0xd9b
github.com/jhump/protoreflect/desc/protoparse.createParseResult(0xc0002b6c30, 0xe, 0xc00038c180, 0xc00039c030, 0xd42f00)
github.com/jhump/protoreflect@v1.7.1-0.20200706205124-8d2d934ce565/desc/protoparse/parser.go:723 +0xba
github.com/jhump/protoreflect/desc/protoparse.parseProto(0xc0002b6c30, 0xe, 0x7f82285a80e0, 0xc00038e080, 0xc00039c030, 0xc0002b6c01, 0xe)
github.com/jhump/protoreflect@v1.7.1-0.20200706205124-8d2d934ce565/desc/protoparse/parser.go:702 +0x16e
github.com/jhump/protoreflect/desc/protoparse.parseProtoFile.func2(0x7f82285a80b8, 0xc00038e080, 0xc0002b6c30, 0xe, 0xc00039c030, 0x1, 0xc00008fb00)
github.com/jhump/protoreflect@v1.7.1-0.20200706205124-8d2d934ce565/desc/protoparse/parser.go:473 +0xba
github.com/jhump/protoreflect/desc/protoparse.parseProtoFile(0xc00039a000, 0xc0002b6c30, 0xe, 0x0, 0xc00039c030, 0x400101, 0xc00008fcc8, 0xc4d950)
github.com/jhump/protoreflect@v1.7.1-0.20200706205124-8d2d934ce565/desc/protoparse/parser.go:474 +0x958
github.com/jhump/protoreflect/desc/protoparse.parseProtoFiles(0xc00039a000, 0xc0002baa80, 0xd, 0xd, 0xc00039c030, 0x101, 0xc00008fcc8, 0x0)
github.com/jhump/protoreflect@v1.7.1-0.20200706205124-8d2d934ce565/desc/protoparse/parser.go:447 +0xb9
github.com/jhump/protoreflect/desc/protoparse.Parser.ParseFiles(0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc00039a000, 0x1, 0xc00038e020, 0x0, ...)
github.com/jhump/protoreflect@v1.7.1-0.20200706205124-8d2d934ce565/desc/protoparse/parser.go:217 +0x211
github.com/bufbuild/buf/internal/buf/bufbuild.getBuildResult(0xd3b620, 0xc0000bb560, 0xc0001d53b0, 0xc0002baa80, 0xd, 0xd, 0x0, 0x0)
github.com/bufbuild/buf/internal/buf/bufbuild/builder.go:179 +0x176
github.com/bufbuild/buf/internal/buf/bufbuild.(*builder).getBuildResults.func1(0xc00030aa80, 0xd3b620, 0xc0000bb560, 0xc0001d53b0, 0xc0002baa80, 0xd, 0xd, 0x203000)
github.com/bufbuild/buf/internal/buf/bufbuild/builder.go:138 +0x73
created by github.com/bufbuild/buf/internal/buf/bufbuild.(*builder).getBuildResults
github.com/bufbuild/buf/internal/buf/bufbuild/builder.go:137 +0x211
```

This seems to indicate that there is a weakness in buf's parser when presented with non-proto files.  It should definitely handle that type of failure better.  Thanks!

**Comment by bufdev** [comment]
Thanks for filing - we actually happened upon this issue a couple days ago, by coincidence. Working on a fix that we will issue soon.

**Comment by aw-was-here** [comment]
Awesome! Thanks!

Since you've seen it live already, buf.yaml doesn't actually help here, correct?

**Comment by bufdev** [comment]
It almost for sure won’t. The bigger thing would. E to rename that .proto Makefile - compilers such as  Golang also use file extensions so it’s not an uncommon pattern :-)

**Comment by bufdev** [comment]
Fixed in #108. You can test it via `go get github.com/bufbuild/buf/cmd/buf@master`, and we will do a release soon. Thanks again!

**Comment by michilu** [comment]
I saw a similar error with only valid proto files via buf@master. How to self check? Do you have some ideas?

```console
$ buf check lint
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1577de7]

goroutine 22 [running]:
github.com/jhump/protoreflect/dynamic.(*Message).findMissingFields(0x0, 0x80, 0x69, 0x1957e20)
        go/pkg/mod/github.com/jhump/protoreflect@v1.7.1-0.20200713144533-3216fce50c70/dynamic/dynamic_message.go:2592 +0x37
github.com/jhump/protoreflect/dynamic.(*Message).validateRecursive(0x0, 0xc00031ca10, 0x6b, 0x6b, 0xc0004310e8)
        go/pkg/mod/github.com/jhump/protoreflect@v1.7.1-0.20200713144533-3216fce50c70/dynamic/dynamic_message.go:2615 +0x43
github.com/jhump/protoreflect/dynamic.(*Message).validateRecursive.func1(0x1944060, 0x0, 0x0, 0x1944060)
        go/pkg/mod/github.com/jhump/protoreflect@v1.7.1-0.20200713144533-3216fce50c70/dynamic/dynamic_message.go:2636 +0x63
github.com/jhump/protoreflect/dynamic.(*Message).validateRecursive(0xc0008187b0, 0xc0001fff90, 0x4a, 0x4a, 0xc000431390)
        go/pkg/mod/github.com/jhump/protoreflect@v1.7.1-0.20200713144533-3216fce50c70/dynamic/dynamic_message.go:2647 +0x4a1
github.com/jhump/protoreflect/dynamic.(*Message).validateRecursive.func1(0x1944060, 0xc0008187b0, 0xc0008187b0, 0x1944060)
        go/pkg/mod/github.com/jhump/protoreflect@v1.7.1-0.20200713144533-3216fce50c70/dynamic/dynamic_message.go:2636 +0x63
github.com/jhump/protoreflect/dynamic.(*Message).validateRecursive(0xc000818780, 0xc000286100, 0x3e, 0x3e, 0xc000431638)
        go/pkg/mod/github.com/jhump/protoreflect@v1.7.1-0.20200713144533-3216fce50c70/dynamic/dynamic_message.go:2657 +0x924
github.com/jhump/protoreflect/dynamic.(*Message).validateRecursive.func1(0x1944060, 0xc000818780, 0xc000818780, 0x1944060)
        go/pkg/mod/github.com/jhump/protoreflect@v1.7.1-0.20200713144533-3216fce50c70/dynamic/dynamic_message.go:2636 +0x63
github.com/jhump/protoreflect/dynamic.(*Message).validateRecursive(0xc0008185d0, 0x0, 0x0, 0xc000140180, 0xc0008185d0)
        go/pkg/mod/github.com/jhump/protoreflect@v1.7.1-0.20200713144533-3216fce50c70/dynamic/dynamic_message.go:2663 +0xadb
github.com/jhump/protoreflect/dynamic.(*Message).ValidateRecursive(...)
        go/pkg/mod/github.com/jhump/protoreflect@v1.7.1-0.20200713144533-3216fce50c70/dynamic/dynamic_message.go:2611
github.com/jhump/protoreflect/desc/protoparse.interpretOptions(0xc000260b40, 0xb1041a0, 0xc0000a6cd0, 0x1944ee0, 0xc00024c240, 0xc0003158a0, 0x2, 0x2, 0x0, 0x0, ...)
        go/pkg/mod/github.com/jhump/protoreflect@v1.7.1-0.20200713144533-3216fce50c70/desc/protoparse/options.go:988 +0xab4
github.com/jhump/protoreflect/desc/protoparse.interpretFileOptions(0xc000260b40, 0x1952400, 0xc0003280e0, 0x12, 0xc00011b2e8)
        go/pkg/mod/github.com/jhump/protoreflect@v1.7.1-0.20200713144533-3216fce50c70/desc/protoparse/options.go:719 +0x3c9
github.com/jhump/protoreflect/desc/protoparse.(*linker).linkFiles(0xc000254dc0, 0xc000254dc0, 0x4, 0x4)
        go/pkg/mod/github.com/jhump/protoreflect@v1.7.1-0.20200713144533-3216fce50c70/desc/protoparse/linker.go:66 +0x1af
github.com/jhump/protoreflect/desc/protoparse.Parser.ParseFiles(0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc000314d70, 0x1, 0xc000349380, 0x0, ...)
        go/pkg/mod/github.com/jhump/protoreflect@v1.7.1-0.20200713144533-3216fce50c70/desc/protoparse/parser.go:225 +0x2cd
github.com/bufbuild/buf/internal/buf/bufbuild.getBuildResult(0x1947ee0, 0xc000133680, 0xc000257400, 0xc0003526c0, 0x4, 0x4, 0x0, 0x0)
        go/pkg/mod/github.com/bufbuild/buf@v0.20.2-0.20200714173906-20a99af2969d/internal/buf/bufbuild/builder.go:179 +0x176
github.com/bufbuild/buf/internal/buf/bufbuild.(*builder).getBuildResults.func1(0xc0001339e0, 0x1947ee0, 0xc000133680, 0xc000257400, 0xc0003526c0, 0x4, 0x4, 0x203000)
        go/pkg/mod/github.com/bufbuild/buf@v0.20.2-0.20200714173906-20a99af2969d/internal/buf/bufbuild/builder.go:138 +0x73
created by github.com/bufbuild/buf/internal/buf/bufbuild.(*builder).getBuildResults
        go/pkg/mod/github.com/bufbuild/buf@v0.20.2-0.20200714173906-20a99af2969d/internal/buf/bufbuild/builder.go:137 +0x211
```

**Comment by johanbrandhorst** [comment]
This is a different crash, we'd be very interested if you could try and pin down exactly the format of the protofile that is causing this to happen, ideally if you could submit the files to us we can take a look, but if that is not possible, a minimal example would be great.

**Comment by michilu** [comment]
@johanbrandhorst I uploaded my codes here. https://github.com/michilu/buf.build-reports/tree/issue-107
Thanks!

**Comment by johanbrandhorst** [comment]
Thanks for that, I managed to reproduce it locally!

**Comment by bufdev** [comment]
Fixed in v0.20.4!
